### PR TITLE
Tests of chat sequence

### PR DESCRIPTION
**Pull request by AlexUstinov**

### CODE REVIEW

File: tests/MLSearch.UnitTests/Indexing/Initializer/InfiniteChatSequenceTests.cs [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+        }
+
+        chats.Verify(
+            ListChangedCall,
+            Times.Exactly((maxChatCount + batchSize - 1) / batchSize)
+        );
+    }
+
+    [Fact]
+    public async Task ThereIsADelayBeforeRetryingInCaseOfAnEmptyBatch()
```
Comment (frolyo) [inline tests/MLSearch.UnitTests/Indexing/Initializer/InfiniteChatSequenceTests.cs]: We don't use articles

File: tests/MLSearch.UnitTests/Indexing/Initializer/InfiniteChatSequenceTests.cs [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+    {
+        var emptyBatchDelay = TimeSpan.FromSeconds(11111);
+
+        var clock = new Mock<IMomentClock>();
+        clock
+            .Setup(x => x.Delay(It.IsAny<TimeSpan>(), It.IsAny<CancellationToken>()))
```
Comment (frolyo) [inline tests/MLSearch.UnitTests/Indexing/Initializer/InfiniteChatSequenceTests.cs:64]: afaik in such case we write the first method call on the same line as variable

File: tests/MLSearch.UnitTests/Indexing/Initializer/InfiniteChatSequenceTests.cs [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+            .Returns(Task.CompletedTask);
+
+        var batchCount = 0;
+        var chats = new Mock<IChatsBackend>();
+        chats
+            .Setup(ListChangedCall)
+            .Returns<long, long, ChatId, int, CancellationToken>(
+                (minVersion, _, _, size, _) => batchCount++ == 0
+                    ? Task.FromResult(new ApiArray<Chat.Chat>([]))
```
Comment (frolyo) [inline tests/MLSearch.UnitTests/Indexing/Initializer/InfiniteChatSequenceTests.cs]: you can use ApiArray.Empty or [] itself

File: tests/MLSearch.UnitTests/Indexing/Initializer/InfiniteChatSequenceTests.cs [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+
+        var sequence = new InfiniteChatSequence(clock.Object, chats.Object, Log) {
+            NoChatsIdleInterval = emptyBatchDelay
+        };
+
+        await sequence.LoadAsync(1, CancellationToken.None).Take(1).EnumerateAll();
+
+        chats.Verify(ListChangedCall, Times.Exactly(2));
+        clock.Verify(
+            x => x.Delay(It.Is<TimeSpan>(ts => ts==emptyBatchDelay), It.IsAny<CancellationToken>()), Times.Once);
```
Comment (frolyo) [inline tests/MLSearch.UnitTests/Indexing/Initializer/InfiniteChatSequenceTests.cs]: missed spaces

File: tests/MLSearch.UnitTests/Indexing/Initializer/InfiniteChatSequenceTests.cs [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+    }
+
+    private static Task<ApiArray<Chat.Chat>> GetNextBatch(long lastVersion, int batchSize)
+    {
+        var batch = new Chat.Chat[batchSize];
+        for (var i = 0; i < batchSize; i++) {
+            var chatId = new ChatId(Generate.Option);
+            batch[i] = new Chat.Chat(chatId, lastVersion + i + 1);
+        }
+        return Task.FromResult(new ApiArray<Chat.Chat>(batch));
```
Comment (frolyo) [inline tests/MLSearch.UnitTests/Indexing/Initializer/InfiniteChatSequenceTests.cs]: batch.ToApiArray()

File: tests/MLSearch.UnitTests/Indexing/Initializer/InfiniteChatSequenceTests.cs [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+    }
+
+    [Fact]
+    public async Task LoadMethodThrowsIfCancellationRequestedBeforeCall()
+    {
+        var sequence = new InfiniteChatSequence(Clock, Chats, Log);
+
+        var cancellationSource = new CancellationTokenSource();
+        await cancellationSource.CancelAsync();
+        var e = await Assert.ThrowsAsync<OperationCanceledException>(async () => {
```
Comment (frolyo) [inline tests/MLSearch.UnitTests/Indexing/Initializer/InfiniteChatSequenceTests.cs:106]: btw, there is an extension which convers Task/ValueTask to Func<Task> and allows to use fluent assertions in a fluent manner:
```
await sequence.LoadAsync(1, cancellationSource.Token).Take(1).EnumerateAll()
  .AsAsyncFunc().Should().ThrowAsync<OperationCancelledException>
```

**Review by alexis-kochetov — APPROVED** [review]

File: tests/MLSearch.UnitTests/Indexing/Initializer/InfiniteChatSequenceTests.cs (hunk 6 above)
Comment (AlexUstinov) [reply tests/MLSearch.UnitTests/Indexing/Initializer/InfiniteChatSequenceTests.cs:106]: I would use this one, but I need the exception instance to run one more assertion in the next statement.

File: tests/MLSearch.UnitTests/Indexing/Initializer/InfiniteChatSequenceTests.cs (hunk 1 above)
Comment (AlexUstinov) [reply tests/MLSearch.UnitTests/Indexing/Initializer/InfiniteChatSequenceTests.cs]: Done

File: tests/MLSearch.UnitTests/Indexing/Initializer/InfiniteChatSequenceTests.cs (hunk 3 above)
Comment (AlexUstinov) [reply tests/MLSearch.UnitTests/Indexing/Initializer/InfiniteChatSequenceTests.cs]: Done

File: tests/MLSearch.UnitTests/Indexing/Initializer/InfiniteChatSequenceTests.cs (hunk 4 above)
Comment (AlexUstinov) [reply tests/MLSearch.UnitTests/Indexing/Initializer/InfiniteChatSequenceTests.cs]: Done

File: tests/MLSearch.UnitTests/Indexing/Initializer/InfiniteChatSequenceTests.cs (hunk 5 above)
Comment (AlexUstinov) [reply tests/MLSearch.UnitTests/Indexing/Initializer/InfiniteChatSequenceTests.cs]: Done

File: tests/MLSearch.UnitTests/Indexing/Initializer/InfiniteChatSequenceTests.cs (hunk 2 above)
Comment (AlexUstinov) [reply tests/MLSearch.UnitTests/Indexing/Initializer/InfiniteChatSequenceTests.cs:64]: It looks uglier in this certain case. I decided to keep this style for now.

**Pull request merged**